geps/gep-1713/index.md

```diff
@@ -0,0 +1,377 @@
+# GEP-1713: Standard Mechanism to Merge Multiple Gateways
+
+* Issue: [#1713](/kubernetes-sigs/gateway-api/issues/1713)
+* Status: Provisional
+
+(See status definitions [here](overview.md#status).)
+
+## tl;dr
+
+The Gateway Resource is a contention point since it is the only place to attach listeners with certificates. We propose a mechanism to allow distinct Gateway resources to be logically merged.
+
+## Goals
+
+- Define a mechanic to merge multiple Gateways (logically)
+- Define a set of acceptable properties that can be merged and their semantics
+
+## Non-Goals
+
+- Apply a Gateway resource onto N distinct gateways (one to many)
+
+## Introduction
+
+Knative generates on demand per-service certificates using HTTP-01 challenges. 
+There can be O(1000) Knative Services in the cluster which means we have O(1000) distinct certificates. 
+Thus updating a single Gateway resource with this many certificates is a contention point and inhibits horizontal scaling of our controllers.
+[Istio Ambient](https://istio.io/v1.15/blog/2022/introducing-ambient-mesh/), similarly, creates a listener per Kubernetes service.
+
+More broadly, large scale gateway users often expose O(1000) domains, but are currently limited by the maximum of 16 `listeners`.
+
+The spec currently has language to indicate implementations `MAY` merge Gateways resources but the mechanic isn't defined.
+https://github.com/kubernetes-sigs/gateway-api/blob/541e9fc2b3c2f62915cb58dc0ee5e43e4096b3e2/apis/v1beta1/gateway_types.go#L76-L78
+
+## API
+
+This proposal has two aspects: configuration for the parent Gateway, and configuration for the child Gateway.
+
+A "parent" Gateway _does not_ reference another Gateway. A parent Gateway MUST explicitly opt into merging.
+This is done with a new field `allowedChildren` in the `spec.infrastructure` of a Gateway.
+If, and only if, they have done so, the same Gateway is also permitted to leave `listeners` empty (currently, there is a `MinItems=1` restriction).
```

Comment:
I'm definitely support of this but its worth pointing out that this does not represent a guarantee that the merged set produces MinItems=1 since there may be no children. Thus we cannot completely state that runtime merging is the exact equivalent of a single merged resource.  

Given that runtime handling of empty lists by controllers was given as a reason for rejecting

https://github.com/kubernetes-sigs/gateway-api/pull/1596
https://github.com/kubernetes-sigs/gateway-api/issues/1817

we may have to reconsider the logic of that decision given the desire to support merging. E.g.

"if controllers want to support merging as per the spec they _must_ tolerate empty listeners even if merging is not used"

It seems like we are making a similar constraint loosening decision around certs for TLS listeners
https://github.com/kubernetes-sigs/gateway-api/pull/2721

Reply:
If I may toss in an opinion here..  I've been watching this proposal because I like the use case of service-based teams and trusting them to manage any needed domains.  It would be ideal if these teams didn't have to disrupt core infrastructure resources (e.g. a top-level 80/443 gateway) before deploying their domain.

Could this be more specific about the way `allowedChildren` should be used?  Could it specify an intent to support regex and/or broader matching, e.g. all gateways in a namespace.?

Reply:
@cornfeedhobo I'd expect `allowedChildren` to look very similar to [AllowedRoutes](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io%2fv1.AllowedRoutes), and specifically the `namespaces` field type [RouteNamespaces](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1.RouteNamespaces), which yes would support matching all gateways one or more namespaces.

Reply:
One thing that just occurred to me when rereading this is that changing `minItems` to 0 for Listeners means that we are moving Listeners from _required_ to _optional_, which is a breaking API change (because fields that you used to be able to assume would always have at least one entry now do not). Strictly, that would mean we'd have to do an API revision to v2, which does not seem likely.

Realistically, it _may_ not be too much of a problem for Go implementations, if they're doing a `range` across `spec.listeners`, since that will execute the code 0 times if the list is empty. But it is a _substantial_ UX change to be able to add Gateways with no Listeners. It's _very_ likely that code will break as a result.

Reply:
I wonder if this could be solved conditionally for example we allow `minItems == 0 `if `spec.infrastructure.allowedChildren` is not nil which implies the Gateway will be a parent. @robscott is that something that is possible with CEL?

> But it is a substantial UX change to be able to add Gateways with no Listeners. It's very likely that code will break as a result.

This doesn't break existing users or implementations. I'd say this is more analogous to supporting a new feature in a newer version of the API.

When an implementation adds 1.2 support they can reject Gateways with no listeners if they do not support gateway merging etc.

Reply:
CEL was the intent here and how we implement "If, and only if, they have done so..."

Reply:
I'm not keen on enforcing the parent/child relationship for this.  It might not be a bad idea to consider a policy or other CRD to record the members of a group that can merge with each other.  Then you are free to design a solution without the restrictions of breaking API changes.